## Tiền xử lý dữ liệu

In [1]:
import polars as pl 
import matplotlib.pyplot as plt

df = pl.scan_csv("output/20250715-ketquathi-ct2018a.csv")

df = df.with_columns([
    pl.col("SOBAODANH").cast(pl.Utf8).str.zfill(8).alias("SOBAODANH"),
    
    *[
    pl.col(col).cast(pl.Float64).alias(col)
        for col in df.collect_schema().names()
        if col not in ["STT", "SOBAODANH", "Mã tỉnh thành"]
    ]
])

df.collect().sample(n=5)

shape: (5, 21)
┌────────┬───────────┬──────┬──────┬───┬──────┬──────┬──────┬───────────────┐
│ STT    ┆ SOBAODANH ┆ Toán ┆ Văn  ┆ … ┆ Đức  ┆ Pháp ┆ Nga  ┆ Mã tỉnh thành │
│ ---    ┆ ---       ┆ ---  ┆ ---  ┆   ┆ ---  ┆ ---  ┆ ---  ┆ ---           │
│ i64    ┆ str       ┆ f64  ┆ f64  ┆   ┆ f64  ┆ f64  ┆ f64  ┆ i64           │
╞════════╪═══════════╪══════╪══════╪═══╪══════╪══════╪══════╪═══════════════╡
│ 483490 ┆ 22011279  ┆ 3.5  ┆ 6.75 ┆ … ┆ null ┆ null ┆ null ┆ 22            │
│ 451806 ┆ 21006036  ┆ 4.35 ┆ 8.5  ┆ … ┆ null ┆ null ┆ null ┆ 21            │
│ 15445  ┆ 01015592  ┆ 1.45 ┆ 6.5  ┆ … ┆ null ┆ null ┆ null ┆ 10            │
│ 543715 ┆ 26009549  ┆ 4.75 ┆ 6.25 ┆ … ┆ null ┆ null ┆ null ┆ 26            │
│ 761537 ┆ 38003546  ┆ 4.0  ┆ 5.5  ┆ … ┆ null ┆ null ┆ null ┆ 38            │
└────────┴───────────┴──────┴──────┴───┴──────┴──────┴──────┴───────────────┘

In [2]:
df.collect_schema().names()

['STT',
 'SOBAODANH',
 'Toán',
 'Văn',
 'Lí',
 'Hóa',
 'Sinh',
 'Tin học',
 'Công nghệ công nghiệp',
 'Công nghệ nông nghiệp',
 'Sử',
 'Địa',
 'Giáo dục kinh tế và pháp luật',
 'Anh',
 'Nhật',
 'Trung',
 'Hàn',
 'Đức',
 'Pháp',
 'Nga',
 'Mã tỉnh thành']

In [3]:
df.describe()

shape: (9, 22)
┌────────────┬───────────────┬───────────┬────────────┬───┬────────────┬────────────┬────────────┬───────────────┐
│ statistic  ┆ STT           ┆ SOBAODANH ┆ Toán       ┆ … ┆ Đức        ┆ Pháp       ┆ Nga        ┆ Mã tỉnh thành │
│ ---        ┆ ---           ┆ ---       ┆ ---        ┆   ┆ ---        ┆ ---        ┆ ---        ┆ ---           │
│ str        ┆ f64           ┆ str       ┆ f64        ┆   ┆ f64        ┆ f64        ┆ f64        ┆ f64           │
╞════════════╪═══════════════╪═══════════╪════════════╪═══╪════════════╪════════════╪════════════╪═══════════════╡
│ count      ┆ 1.131136e6    ┆ 1131136   ┆ 1.131136e6 ┆ … ┆ 150.0      ┆ 408.0      ┆ 60.0       ┆ 1.131136e6    │
│ null_count ┆ 0.0           ┆ 0         ┆ 0.0        ┆ … ┆ 1.130986e6 ┆ 1.130728e6 ┆ 1.131076e6 ┆ 0.0           │
│ mean       ┆ 565568.5      ┆ null      ┆ 4.762109   ┆ … ┆ 7.816667   ┆ 8.376838   ┆ 8.941667   ┆ 32.572542     │
│ std        ┆ 326530.981383 ┆ null      ┆ 1.703135   ┆ … ┆ 1.641448   ┆ 1.46792    ┆ 1.156564   ┆ 16.986754     │
│ min        ┆ 1.0           ┆ 01000001  ┆ 0.0        ┆ … ┆ 3.5        ┆ 3.25       ┆ 5.25       ┆ 10.0          │
│ 25%        ┆ 282785.0      ┆ null      ┆ 3.5        ┆ … ┆ 6.75       ┆ 7.75       ┆ 8.75       ┆ 20.0          │
│ 50%        ┆ 565569.0      ┆ null      ┆ 4.6        ┆ … ┆ 8.25       ┆ 8.75       ┆ 9.25       ┆ 29.0          │
│ 75%        ┆ 848352.0      ┆ null      ┆ 5.85       ┆ … ┆ 9.0        ┆ 9.5        ┆ 9.75       ┆ 47.0          │
│ max        ┆ 1.131136e6    ┆ 64007242  ┆ 10.0       ┆ … ┆ 10.0       ┆ 10.0       ┆ 10.0       ┆ 90.0          │
└────────────┴───────────────┴───────────┴────────────┴───┴────────────┴────────────┴────────────┴───────────────┘

In [4]:
necessary_df_bool = df.drop(["STT", "SOBAODANH", "Toán", "Văn", "Mã tỉnh thành"])
    
necessary_df_bool = necessary_df_bool.select(
    [pl.col(c).is_not_null().alias(c) for c in necessary_df_bool.collect_schema().names()]
).collect().to_pandas()

necessary_df_bool.head()

Lí    Hóa   Sinh  Tin học  Công nghệ công nghiệp  Công nghệ nông nghiệp  \
0  False   True   True    False                  False                  False   
1   True   True  False    False                  False                  False   
2   True  False  False    False                  False                  False   
3   True   True  False    False                  False                  False   
4  False  False  False    False                  False                  False   

      Sử    Địa  Giáo dục kinh tế và pháp luật    Anh   Nhật  Trung    Hàn  \
0  False  False                          False  False  False  False  False   
1  False  False                          False  False  False  False  False   
2  False  False                          False  False  False  False  False   
3  False  False                          False  False  False  False  False   
4   True   True                          False  False  False  False  False   

     Đức   Pháp    Nga  
0  False  False  False  
1  False  False  False  
2  False  False  False  
3  False  False  False  
4  False  False  False

## Luật kết hợp (Apriori)

In [5]:
from mlxtend.frequent_patterns import apriori, association_rules, fpgrowth
import matplotlib.pyplot as plt

In [20]:
combine = apriori(
    necessary_df_bool, 
    min_support = 0.01, 
    use_colnames = True
)

rules = association_rules(
    combine, 
    metric="confidence", 
    min_threshold=0.25
)

In [21]:
rules[['antecedents', 'consequents', 'support', 'confidence', 'lift', 'leverage']].round(2)

antecedents consequents  support  confidence  lift  \
0                            (Hóa)        (Lí)     0.14        0.68  2.20   
1                             (Lí)       (Hóa)     0.14        0.47  2.20   
2                             (Lí)       (Anh)     0.13        0.42  1.36   
3                            (Anh)        (Lí)     0.13        0.42  1.36   
4                           (Sinh)       (Hóa)     0.04        0.64  3.03   
5                             (Sử)       (Địa)     0.26        0.62  1.47   
6                            (Địa)        (Sử)     0.26        0.62  1.47   
7  (Giáo dục kinh tế và pháp luật)        (Sử)     0.06        0.28  0.65   
8  (Giáo dục kinh tế và pháp luật)       (Địa)     0.10        0.44  1.05   

   leverage  
0      0.08  
1      0.08  
2      0.03  
3      0.03  
4      0.03  
5      0.08  
6      0.08  
7     -0.03  
8      0.00

## Luật kết hợp (FPGrowth)

In [17]:
combine = fpgrowth(
    necessary_df_bool, 
    min_support = 0.01, 
    use_colnames = True
)

rules = association_rules(
    combine, 
    metric="confidence", 
    min_threshold=0.25
)

In [18]:
print(rules[['antecedents', 'consequents', 'support', 'confidence', 'lift', 'leverage']].round(2))

                       antecedents consequents  support  confidence  lift  \
0                            (Hóa)        (Lí)     0.14        0.68  2.20   
1                             (Lí)       (Hóa)     0.14        0.47  2.20   
2                           (Sinh)       (Hóa)     0.04        0.64  3.03   
3                             (Lí)       (Anh)     0.13        0.42  1.36   
4                            (Anh)        (Lí)     0.13        0.42  1.36   
5                             (Sử)       (Địa)     0.26        0.62  1.47   
6                            (Địa)        (Sử)     0.26        0.62  1.47   
7  (Giáo dục kinh tế và pháp luật)        (Sử)     0.06        0.28  0.65   
8  (Giáo dục kinh tế và pháp luật)       (Địa)     0.10        0.44  1.05   

   leverage  
0      0.08  
1      0.08  
2      0.03  
3      0.03  
4      0.03  
5      0.08  
6      0.08  
7     -0.03  
8      0.00  


## Kết luận

1. **Nhóm Tự nhiên**

- **Hóa ↔ Lí**
    - Support ~14% → khá nhiều thí sinh chọn cả hai.
    - Confidence Hóa → Lí = 68% (rất cao).
    - Lift = 2.2 → mối liên hệ mạnh gấp đôi so với ngẫu nhiên.
    - 👉 Đây là cặp môn đi cùng nhau rõ rệt, phản ánh truyền thống khối A.

- **Sinh → Hóa**
    - Confidence = 64%, Lift = 3.0 → cực kỳ mạnh.
    - 👉 Thí sinh chọn Sinh gần như chắc chắn chọn Hóa, phản ánh khối B (Toán–Hóa–Sinh).
    - Ngược lại, Hóa → Sinh không mạnh bằng, vì Hóa còn gắn với khối A.

2. **Nhóm Xã hội**

- **Sử ↔ Địa**
    - Support ~26%, Confidence ~62%, Lift ~1.47 → mối liên hệ khá rõ.
    - 👉 Đây là cặp môn đặc trưng của khối C (Văn–Sử–Địa).

3. **Các mối liên hệ khác**

- **Anh ↔ Lí**
    - Confidence ~42%, Lift ~1.36 → có xu hướng, nhưng không mạnh bằng các cặp truyền thống.
    - 👉 Có thể phản ánh một số thí sinh chọn kết hợp khối D với môn tự nhiên.

- **GD Kinh tế & Pháp luật → Địa**
    - Lift ≈ 1.05
    - 👉 gần như ngẫu nhiên, không đáng chú ý.


==> **KẾT LUẬN**

- Đáng chú ý nhất: Nhóm Tự nhiên (Hóa–Lí–Sinh) và Nhóm Xã hội (Sử–Địa).
- Một chiều đặc trưng: Sinh → Hóa.
- Hai chiều mạnh: Hóa ↔ Lí, Sử ↔ Địa.
- Không đáng chú ý: GD Kinh tế & Pháp luật → Địa.